cmake/std/atdm/ats2/trilinos_jsrun

```diff
@@ -0,0 +1,119 @@
+#!/bin/bash
+
+DEBUG_SCRIPT="${DEBUG_SCRIPT-:0}"
+ECHO_CMD="${ECHO_CMD=:0}"
+np_is_one="false"
+orig_args=("$@")
+args=("$@")
+
+function debug_print {
+  if [ "$DEBUG_SCRIPT" == "1" ]; then
+    msg="$@"
+    1>&2 echo "$msg"
+  fi
+}
+
+for i in "${!args[@]}";
+do
+  arg=${args[${i}]}
+
+  debug_print "Processing arg=$arg"
+
+  # handle the case of seeing the np argument
+  # lets assume -p ## or --np=##
+  if [[ "$arg" =~ ^-p[0-9]+$ ]]; then
+    # np is just the 2nd arg split on 'p'
+    my_np=$(echo "$arg" | cut -f2 -d'p');
+    debug_print "  Case: -p##, with ##=$my_np"
+    if [ "$my_np" == "1" ]; then
+      np_is_one="true";
+      debug_print "  Case: -p##, with ##=$my_np and np=1"
+    fi
+    # we are finished
+    break;
+  # next, -p ##
+  elif [ "$arg" == "-p" ]; then
+    ip1=$(($i+1));
+    my_np=${args[${ip1}]}
+    debug_print "  Case: -p ##, with ##=$my_np"
+    if [ "$my_np" == "1" ]; then
+      np_is_one="true";
+      debug_print "  Case: -p ##, with ##=$my_np and np=1"
+    fi
+    # we are finished
+    break;
+  # next, --np=##
+  elif [[ "$arg" =~ ^--np=[0-9]+$ ]]; then
+    # np is just the 2nd arg split on '='
+    my_np=$(echo "$arg" | cut -f2 -d'=');
+    debug_print "  Case: --np=##, with ##=$my_np"
+    if [ "$my_np" == "1" ]; then
+      np_is_one="true";
+      debug_print "  Case: --np=##, with ##=$my_np and np=1"
+    fi
+    # we are finished
+    break;
+  fi
+done
+
+if [ "$np_is_one" == "true" ]; then
+  added_disable="false"
+  debug_print "NP=1 adding disable_gpu_hooks"
+  # look for an '-M' arg and append disable hooks to the next argument
+  for i in "${!args[@]}";
+  do
+    arg=${args[${i}]}
+    debug_print "arg=$arg"
+    if [ "$arg" == "-M" ]; then
+      debug_print "arg=$arg, found -M, prepending -disable_gpu_hooks"
+      ip1=$(($i+1));
+      # single quotes matter, as the argument to -M needs to be all args originally quoted to it
+      args[${ip1}]="'-disable_gpu_hooks ${args[${ip1}]}'"
+      debug_print "updated arg = ${args[${ip1}]}"
+      added_disable="true"
+      break
+    fi
+  done
+
+  # did not find -M, so we add it
+  if [ "$added_disable" == "false" ]; then
+    debug_print "NP=1, but we didn't find a -M argument, so adding -M and disabling the gpu_hooks"
+    # prepend the argument
+    args=("-M -disable_gpu_hooks" "${args[@]}")
+  fi
+fi
+
+# I think if np=1 and disable_gpu_hooks is passed, then you aren't supposed to use -gpu
```

Comment:
@jjellio,

- Thinking about this more now, I am confused. I don't have a very good understanding of Trilinos, MPI, etc. I'm also not sure exactly how test commands are configured in Trilinos. For the ATDM build environment, it sets "mpiexec flags" and an "mpiexec" program for a given system, like vortex, for example. I believe these "mpiexec" options apply to all tests that are enabled at configure time and that individual tests or packages do not set their own "mpiexec" options. @bartlettroscoe, can you confirm?

- Why is `/home/projects/ppc64le-pwr9-nvidia/openmpi/2.1.2/gcc/7.2.0/cuda/9.2.88/bin/mpiexec` able to handle executing both mpi based tests and cuda+mpi based tests on waterman? In looking at the waterman Jenkins projects and test commands, waterman just runs a cuda aware mpiexec and doesn't parse mpi arguments and hack them. Does `/home/projects/ppc64le-pwr9-nvidia/openmpi/2.1.2/gcc/7.2.0/cuda/9.2.88/bin/mpiexec` look at the application binary and determine which HW devices that binary is targeting?

I am concerned that #6718 and #6691 are attempting to work around an issue that should be fixed in spectrum MPI.

I think there are actually 3 issues that should be addressed:
1. The way we currently run tests doesn't exercise both paths affected by setting `TPETRA_ASSUME_CUDA_AWARE_MPI` to either `0` or `1`.
2. Spectrum MPI needs to be updated to properly execute application binaries that target either GPU, HOST, or  HOST+GPU.
3. Until 2. is addressed, work around it with #6718 and #6691.

Thoughts?

Reply:
MPI on waterman/ride/white is defaulting to enable cuda-aware. On Vortex, you shouldn't use mpiexec, instead you need to use jsrun (similar to using srun on Mutrino).

To be honest, many of the issues we've had are because the MPI on waterman/white are *not* making it clear which code path they are using. This explanation is probably easier in person.

Tpetra supports a runtime option to decide whether to use device buffers w/MPI, or to only use host-side buffers with MPI. Tpetra configures the **default** setting for this runtime parameter at *build time*.   If CMake determined that MPI is cuda-aware, then Tpetra will assume it is cuda aware and pass device buffers.  You can change this behavior by exporting `TPETRA_ASSUME_CUDA_AWARE_MPI`, and at Tpetra::initialize it will change the default.

What stinks, is that running configure on the login node may report that MPI is not cuda aware (which was the case on LLNL Lassen).  Yet, on the compute nodes, MPI would report it is cuda-aware.  This leads to confusion over which implementation of MPI we are using.

The implementation of MPI is (cuda-aware or not), is basically like using two different MPI's from a performance/speed perspective.  The flaw in general is that these options are **runtime**, but they require independent code paths inside our own code.  This causes the problems we have right now. Tpetra must implement logic to copy devices buffers to the host if you run on a machine where you can't or don't want to use device buffers w/MPI (but you do want to use the device for compute). 

The end result is that users and developers never had to worry about MPI flags aside from performance tuning (binding things right, choosing your 'np', etc...). With Device specific MPI, we now have MPI implementation details leaking into logic inside libraries like Tpetra. Hence, people can choose (or not set MPI flags), and Tpetra tries to make the library functional <- because some platforms historically had devices but no MPI was allowed with device buffers.

Our own test beds made a switch to device-aware MPI several years ago, but unfortunately I don't think we didn't realize how impactful that was.  For years, our codes have been CI tested with a very specific code path. Now, the fact there are multiple code paths has been thrust into the spot light.  By default, Vortex *use* to report that MPI was not cuda-aware, a very different thing than White (where it would report aware).

Spectrum does support everything for (2).  See my comment in the WIP issue

Reply:
> MPI on waterman/ride/white is defaulting to enable cuda-aware. On Vortex, you shouldn't use mpiexec, instead you need to use jsrun (similar to using srun on Mutrino).
> 
> To be honest, many of the issues we've had are because the MPI on waterman/white are _not_ making it clear which code path they are using. This explanation is probably easier in person.

How is `/home/projects/ppc64le-pwr9-nvidia/openmpi/2.1.2/gcc/7.2.0/cuda/9.2.88/bin/mpiexec` able to handle executing both mpi based tests and cuda+mpi based tests on waterman? I mean, there is no need to pass `-gpu` or `-disable_gpu_hooks`, `/home/projects/ppc64le-pwr9-nvidia/openmpi/2.1.2/gcc/7.2.0/cuda/9.2.88/bin/mpiexec` figures it out, right?

Reply:
No it doesn't figure anything out.

`-M -gpu` is enabling device pointers w/MPI.  Waterman default to this behavior.   With that feature enabled, you get a superset of support. MPI will accept both host and device pointers... so adding `-M -gpu` tends to 'fix' things.... but in reality is isn't fixing anything. It is masking the fact that your code is using MPI in an unpredictable way.  Since Tpetra/Trilinos supports the use case that you can use 'accelerators' independent of whether MPI supports their memory address, then it is important that the library enforce that it passes MPI buffers in a safe way.

The 'feature' of having device memory accessible to MPI can result in a 40x hit in terms of network performance, which is why apps may choose not to use it.  It has also been buggy. Hence, we want to test the case of using Cuda with MPI+Cuda buffers.